File: src/main/java/io/routr/core/LogsHandler.java
package io.routr.core;

import org.eclipse.jetty.websocket.api.*;
import org.eclipse.jetty.websocket.api.annotations.*;
import java.io.*;
import java.util.*;
import java.util.concurrent.*;
import java.lang.System;

/**
 * @author Pedro Sanders
 * @since v1
 */
 @WebSocket
 public class LogsHandler {

     // Store sessions if you want to, for example, broadcast a message to all users
     private static final Queue<Session> sessions = new ConcurrentLinkedQueue<>();

     @OnWebSocketConnect
     public void connected(Session session) {
         sessions.add(session);
     }

     @OnWebSocketClose
     public void closed(Session session, int statusCode, String reason) {
         sessions.remove(session);
     }

     @OnWebSocketMessage
     public void message(Session session, String message) throws IOException, InterruptedException{
        BufferedReader br = null;
        try {
            String base = System.getenv("ROUTR_DATA") != null
              ? System.getenv("ROUTR_DATA")
              : ".";
            File file = new File(base + "/logs/routr.log");
            br = new BufferedReader( new FileReader(file) );
            while (true) {
              String line = br.readLine();
              if (line == null) { // end of file, start polling
                  Thread.sleep(5 * 1000);
              } else {
                  session.getRemote().sendString(line);
              }
            }
        } finally {
          if (br != null) br.close();
        }
     }

 }

File: src/main/kotlin/io/routr/core/LogsHandler.kt
package io.routr.core

import org.eclipse.jetty.websocket.api.Session
import org.eclipse.jetty.websocket.api.annotations.OnWebSocketClose
import org.eclipse.jetty.websocket.api.annotations.OnWebSocketConnect
import org.eclipse.jetty.websocket.api.annotations.OnWebSocketMessage
import org.eclipse.jetty.websocket.api.annotations.WebSocket
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue

/**
 * @author Pedro Sanders
 * @since v1
 */
@WebSocket
class LogsHandler {
    @OnWebSocketConnect
    fun connected(session: Session) {
        sessions.add(session)
    }

    @OnWebSocketClose
    fun closed(session: Session, statusCode: Int, reason: String) {
        sessions.remove(session)
    }

    @OnWebSocketMessage
    @Throws(IOException::class, InterruptedException::class)
    fun message(session: Session, message: String) {
        var br: BufferedReader? = null
        try {
            val base = if (System.getenv("DATA") != null) System.getenv("DATA") else "."
            val file = File("$base/logs/routr.log")
            br = BufferedReader(FileReader(file))
            while (true) {
                val line = br.readLine()
                if (line == null) { // end of file, start polling
                    Thread.sleep(5 * 1000.toLong())
                } else {
                    session.remote.sendString(line)
                }
            }
        } finally {
            br?.close()
        }
    }

    companion object {
        // Store sessions if you want to, for example, broadcast a message to all users
        private val sessions: Queue<Session> = ConcurrentLinkedQueue()
    }
}